gulpfile.js

```diff
@@ -124,28 +199,203 @@ gulp.task("default", function (callback) {
+            bootMethod(function () {
+                iOSPlatform.emulatorStart();
+                return checkEmulatorReadyLooper(iOSPlatform);
+            });
+        }
+        
+        // This needs to be done so that the task will exit.
+        // The command that creates the Android emulator persists with the life of the emulator and hangs this process unless we force it to quit.
+        gulp.doneCallback = function (err) {
+            process.exit(err ? 1 : 0);
+        }
+    }
+    
+    if (!ios) startEmulatorsInternal();
+}
 
-gulp.task("test-ios-wkwebview", function (callback) {
-    var command = "mocha";
-    var args = ["./bin/test", "--mockserver", "http://127.0.0.1:3000", "--platform", "ios", "--use-wkwebview", "true"];
-    executeCommand(command, args, callback);
-});
+// procedurally generate tasks for every possible testing configuration
+var cleanSuffix = "-clean";
+var fastSuffix = "-fast";
 
-gulp.task("test-android", function (callback) {
-    var command = "mocha";
-    var args = ["./bin/test", "--mockserver", "http://10.0.2.2:3000", "--platform", "android"];
-    executeCommand(command, args, callback);
-});
+// generate tasks for starting emulators
+function generateEmulatorTasks(taskName, android, ios) {
+    gulp.task(taskName, function (callback) {
+        startEmulators(callback, false, android, ios);
+    });
+    
+    gulp.task(taskName + cleanSuffix, function (callback) {
+        startEmulators(callback, true, android, ios);
+    });
+}
 
-gulp.task("test", function (callback) {
-    runSequence("default", "test-android", "test-ios", callback);
-});
\ No newline at end of file
+function getEmulatorTaskNameSuffix(android, ios) {
+    if (!!android === !!ios) return "";
+    else if (android) return "-android";
+    else return "-ios";
+}
+
+var emulatorTaskNamePrefix = "emulator";
+for (var android = 0; android < 2; android++) {
+    for (var ios = 0; ios < 2; ios++) {
+        if (!android && !ios) continue;
+        generateEmulatorTasks(emulatorTaskNamePrefix + getEmulatorTaskNameSuffix(android, ios), android, ios);
+    }
+}
+                
+function generateTestTasks(taskName, options) {
+    gulp.task(taskName + "-fast", function (callback) {
+        runTests(callback, options);
+    });
+    
+    var emulatorTaskName = emulatorTaskNamePrefix + getEmulatorTaskNameSuffix(options.android, options.ios);
+    
+    gulp.task(taskName, function (callback) {
+        runSequence("default", emulatorTaskName, taskName + fastSuffix, callback);
+    });
+    
+    gulp.task(taskName + "-clean", function (callback) {
+        runSequence("default", emulatorTaskName + cleanSuffix, taskName + fastSuffix, callback);
+    });
+}
+
+// procedurally generate tasks for every possible testing configuration
+var taskNamePrefix = "test";
+for (var android = 0; android < 2; android++) {
+    // 0 = don't run android tests
+    // 1 = run android tests
+    for (var ios = 0; ios < 4; ios++) {
```

Comment:
That means that we not only have a design issue, but also a stability issue and we should fix it. This argument is even stronger given that you mentioned you want to use the same framework for react native. If the test framework or the tests themselves are not stable, enabling continuous integration will not have benefits. It will only make things worse, and waste developer time on test failure investigations. My understanding was that the #96 refactor was intended to make the tests [flow better](https://github.com/Microsoft/cordova-plugin-code-push/pull/96).